src/main/scala-2/chisel3/simulator/scalatest/WithTestingDirectory.scala

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.simulator.scalatest
+
+import chisel3.simulator.HasTestingDirectory
+import java.nio.file.{FileSystems, Path}
+import org.scalatest.TestSuite
+import scala.util.DynamicVariable
+
+/** A mix-in for a Scalatest test suite that will setup the output directory for
+  * you.  E.g., this will create output directories for your tests like the
+  * following:
+  *
+  * {{{
+  * <buildDir>
+  * └── <suite-name>
+  *     └── <scope-1-name>
+  *         └── ...
+  *             └── <scope-n-name>
+  *                 ├── <test-1-name>
+  *                 ├── ...
+  *                 └── <test-n-name>
+  * }}}
+  *
+  * You may change the `buildDir` by overridding a method of the same name in
+  * this trait.
+  *
+  */
+trait WithTestingDirectory { self: TestSuite =>
+
+  /** Return the name of the root test directory.
+    *
+    * For different behavior, please override this in your test suite.
+    */
+  def buildDir: String = "test_run_dir"
+
+  // Assemble all the directories that should be created for this test.  This is
+  // done by examining the test (via a fixture) and then setting a dynamic
+  // variable for that test.  The exacct structure of the test is extracted,
+  // including any nesting (Scalatest scopes) in the test.
+  private val testName: DynamicVariable[List[String]] = new DynamicVariable[List[String]](Nil)
+  override def withFixture(test: NoArgTest) = {
+    testName.withValue(test.scopes.toList :+ test.text) {
+      test()
+    }
+  }
+
+  /** Implementaton of [[HasTestingDirectory]] which sets up the test directory
+    * for you based on settings which make sense in Scalatest.
+    */
+  final implicit def implementation: HasTestingDirectory = new HasTestingDirectory {
+
+    /** Return the test name with some sanitization applied. */
+    final def getTestName = testName.value.map(_.replaceAll(" ", "_").replaceAll("\\W+", ""))
+
+    override def getDirectory(testClassName: String): Path = FileSystems
+      .getDefault()
+      .getPath(buildDir, self.suiteName +: getTestName: _*)
+
+  }
+
+}
```

Comment:
This is something you won't be able to test in our normal flow, but I'd like to see an enhanced error message for what happens if a user tries to use this but forgets to add ScalaTest as a dependency. You can experiment by publishing this locally and trying to call it from Scala-CLI or just making an extra build unit--but in either case try to extend this trait and use it but forget to add Scalatest as a dependency.

Reply:
Basically I think we'll want to catch an Linkage exception or whatever somewhere and just add some more context suggesting the user add the dependency on ScalaTest.

Reply:
(Summary of an offline discussion investigating this:) I looked at what the error would look like if you don't include scalatest as a dependency. Luckily, the fact that this has a self-type on `org.scalatest.TestSuite` means that you will get a compile-time error if you don't try to bring `TestSuite` into scope. If you then try to clear the self-type without providing scalatest as a dependency, you'll get a standard "scalatest is not a member of package org".

tl;dr: the self-type means that other errors will come up before you can hit this. 🎊